xlml/apis/task.py

```diff
@@ -38,6 +41,72 @@ def run() -> DAGNode:
     ...
 
 
+@dataclasses.dataclass
+class DockerBuildTask(BaseTask):
+  """This is a class to set up a Docker build task for XPK workloads.
+
+  XPK runs on Kubernetes, which requires a docker image to execute the workload.
+  To ensure freshness of the images and reduce unnecessary setup on workers, a
+  single BashOperator runs the docker build to prepare the image for execution,
+  then all workers will use the resulting image for their containers.
+
+  The resulting image can be shared across multiple tasks within the DAG, and
+  the unique image name is accessible as the output of the DockerBuildTask.
+
+  Attributes:
+    build_dir: Relative path within the repository to run the build. This path
+        must contain a Dockerfile or cloudbuild.yaml.
+    image_name: The base image name to build. The timestamp of the build will
+        be used to tag the image.
+    custom_build: When set, run the build using cloudbuild.yaml. The
+        ``_IMAGE_NAME`` substitution will be the tagged docker image which will
+        be the output of this AirFlow task. Defaults to False, meaning the build
+        is a direct docker build.
+    task_id: Task ID to use for the operator. Defaults to
+        ``DockerBuildTask.TASK_ID``.
+  """
+
+  build_dir: str
+  image_name: str
+  custom_build: bool = False
+  task_id: Optional[str] = None
+
+  # Path on the Airflow worker which contains the Airflow GCS bucket contents.
+  GCS_PATH = "/home/airflow/gcs/dags"
+  DEFAULT_TASK_ID = "docker_image_build"
+  IMAGE_NAME_SUBSTITUTION = "_IMAGE_NAME"
+  BUILD_SCRIPT = """
+  set -xe
+  cd {build_path}
+  gcloud builds submit --machine-type=e2-highcpu-32 {build_flags} .
+  # The last line of output is pushed to XCom as the task output.
+  echo '{image_tag}'
+  """
+
+  def run(self) -> DAGNode:
```

Comment:
What's the reason for making this a `Task`? BashOperator is itself a DAG node, so you can string that together with other test `Task`s as needed. Or, you could inherit from the base XPK task and override the `run` and/or `provision` methods

Reply:
We want to build a single Docker image and share it across the DAG, so my intent was to keep it separate from the other run/provision workflows. Do you think `Task` isn't the right abstraction here?

Reply:
I think of `Task` as an encapsulation of multiple common steps strung together, which spits out a `DAGNode` so it can interoperate with the rest of airflow. Since you have just a single operator here anyway, I don't think it benefits from being wrapped in a `Task`. 

Since you are doing such a simple call to the GCB API, have you looked at the `google-cloud-build` client library? You could wrap this call in a `@task` [^1]. I am obligated to recommend against any form of log parsing from the `gcloud` CLI https://cloud.google.com/python/docs/reference/cloudbuild/latest

[^1]: Either @RissyRan and I probably made a mistake calling `Task`s "tasks" when they're largely made of `@task`s

Reply:
Gotcha, that makes sense, I'll pull it up into Python.

Reply:
Client library will be a better choice whenever applies. Also, agreed that we called the `Task` with a group of tasks that should be named as task group. Thanks @will-cromar and @jonb377 for this feature!